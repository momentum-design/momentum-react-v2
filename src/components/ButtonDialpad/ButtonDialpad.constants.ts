const CLASS_PREFIX = 'md-button-dialpad';

const STYLE = {
  primaryText: `${CLASS_PREFIX}-primary-text`,
  secondaryText: `${CLASS_PREFIX}-secondary-text`,
  wrapper: `${CLASS_PREFIX}-wrapper`,
  innerWrapper: `${CLASS_PREFIX}-inner-wrapper`,
};

export { CLASS_PREFIX, STYLE };
