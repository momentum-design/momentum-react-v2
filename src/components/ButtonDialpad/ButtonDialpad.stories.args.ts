import {
  commonHTMLAttributes,
  commonMdcButton,
  commonStyles,
} from '../../storybook/helper.stories.argtypes';

export default {
  ...commonStyles,
  ...commonHTMLAttributes,
  ...commonMdcButton,
  primaryText: {
    description: 'Provides the primary text for this `<ButtonDialPad />`.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  secondaryText: {
    description: 'Provides the prsecondaryimary text for this `<ButtonDialPad />`.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
};
