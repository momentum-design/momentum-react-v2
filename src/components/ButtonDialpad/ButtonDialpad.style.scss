.md-button-dialpad-wrapper {
  /* Sizing and spacing. */
  align-items: center;
  border-radius: 100vh;
  height: 4rem;
  width: 4rem;

  > .md-button-dialpad-inner-wrapper {
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;

    > .md-button-dialpad-primary-text {
      font-size: 1.625rem;
      color: var(--mds-color-theme-text-primary-normal);
    }

    > .md-button-dialpad-secondary-text {
      font-size: 0.75rem;
      color: var(--mds-color-theme-text-secondary-normal);
      margin-top: 0.2em;
      height: 1em;
    }

    > * {
      align-items: center;
      display: flex;
      height: 1em;
      justify-content: center;
    }
  }
}
