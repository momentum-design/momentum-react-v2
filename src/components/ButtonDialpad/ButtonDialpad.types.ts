import type { MdcButtonProps } from '../../types';
export interface Props extends Omit<MdcButtonProps, 'size' | 'variant'> {
  /**
   * Primary text to display above the secondary text.
   */
  primaryText?: string;

  /**
   * Secondary text to display below the primary text.
   */
  secondaryText?: string;
}
