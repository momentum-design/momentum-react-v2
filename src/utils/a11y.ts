import { AriaAttributes, useEffect } from 'react';
import { RequireOneOf } from './types';

export type AriaLabelRequired = RequireOneOf<AriaAttributes, ['aria-label', 'aria-labelledby']>;

/**
 * Check if `aria-label` or `aria-labelledby` are defined and are truthy
 *
 * @param componentName The name of the component
 * @param props The object of props passed to the component
 */
export function useCheckAriaLabel(
  componentName: string,
  props: { 'aria-label'?: string; 'aria-labelledby'?: string }
): void {
  useEffect(() => {
    if (!props['aria-label'] && !props['aria-labelledby']) {
      console.warn(`MRV2: ${componentName} requires aria-label or aria-labelledby.`);
    }
  }, [componentName, props]);
}
