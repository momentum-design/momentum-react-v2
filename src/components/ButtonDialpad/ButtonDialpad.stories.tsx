import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import AriaButtonDocs from '../../storybook/docs.stories.aria-button.mdx';
import HTMLGlobalAttributes from '../../storybook/docs.stories.common-html-attributes.mdx';

import ButtonDialpad, { ButtonDialpadProps } from './';
import argTypes from './ButtonDialpad.stories.args';
import Documentation from './ButtonDialpad.stories.docs.mdx';

export default {
  title: 'Momentum UI/ButtonDialpad',
  component: ButtonDialpad,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs, HTMLGlobalAttributes, AriaButtonDocs),
    },
  },
  args: {
    primaryText: '1',
    secondaryText: 'ABC',
  },
};

const Example = Template<ButtonDialpadProps>(ButtonDialpad).bind({});

Example.argTypes = { ...argTypes };

const States = MultiTemplate<ButtonDialpadProps>(ButtonDialpad).bind({});

States.parameters = {
  variants: [{}, { disabled: true }],
};

States.argTypes = { ...argTypes };
delete States.argTypes.disabled;

const Common = MultiTemplate<ButtonDialpadProps>(ButtonDialpad).bind({});

Common.parameters = {
  variants: [
    { primaryText: '1' },
    { primaryText: '2', secondaryText: 'ABC' },
    { primaryText: '3', secondaryText: 'DEF' },
    { primaryText: '4', secondaryText: 'GHI' },
    { primaryText: '5', secondaryText: 'JKL' },
    { primaryText: '6', secondaryText: 'MNO' },
    { primaryText: '7', secondaryText: 'PQRS' },
    { primaryText: '8', secondaryText: 'TUV' },
    { primaryText: '9', secondaryText: 'WXYZ' },
    { primaryText: '0' },
    { primaryText: '*' },
    { primaryText: '#' },
  ],
};

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;
delete Common.argTypes.primaryText;
delete Common.argTypes.secondaryText;

export { Example, States, Common };
