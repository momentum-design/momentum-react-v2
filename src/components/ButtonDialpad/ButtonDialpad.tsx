import React, { forwardRef, RefObject } from 'react';
import { Button as MdcButton } from '@momentum-design/components/dist/react';

import { STYLE } from './ButtonDialpad.constants';
import './ButtonDialpad.style.scss';
import classnames from 'classnames';

import type { Button } from '@momentum-design/components';
import type { Props } from './ButtonDialpad.types';

const ButtonDialpad = forwardRef((props: Props, providedRef: RefObject<Button>) => {
  const { className, ...otherProps } = props;

  return (
    <MdcButton
      className={classnames(STYLE.wrapper, className)}
      ref={providedRef}
      // size 40 is maximum size for MdcButton, setting it here and overriding
      // size in scss file to 64px
      size={40}
      variant="tertiary"
      {...otherProps}
    >
      <div className={STYLE.innerWrapper}>
        <div className={STYLE.primaryText}>{props.primaryText}</div>
        <div className={STYLE.secondaryText}>{props.secondaryText}</div>
      </div>
    </MdcButton>
  );
});

ButtonDialpad.displayName = 'ButtonDialpad';

export default ButtonDialpad;
