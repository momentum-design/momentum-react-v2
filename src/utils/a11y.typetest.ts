import { AriaLabelRequired } from './a11y';
import { Equal, Expect } from './typetest.util';

// NOTE: STATIC TYPE TESTS RUN ON LINT
// eslint-disable-next-line @typescript-eslint/no-unused-vars
type cases = [
  // Check that the AriaLabelRequired type is correct
  Expect<Equal<AriaLabelRequired, { 'aria-label': string } | { 'aria-labelledby': string }>>
];
