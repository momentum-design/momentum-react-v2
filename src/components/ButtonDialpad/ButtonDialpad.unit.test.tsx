import React from 'react';
import { renderWithWebComponent } from '../../../test/utils';
import { ButtonDialpad } from '@momentum-ui/react-collaboration';
import userEvent from '@testing-library/user-event';
import { STYLE } from './ButtonDialpad.constants';
import { screen } from '@testing-library/react';

describe('<ButtonPill />', () => {
  const setup = async (component: React.ReactElement) => {
    return renderWithWebComponent(component);
  };

  describe('snapshot', () => {
    it('should match snapshot', async () => {
      expect.assertions(1);

      const { container } = await setup(<ButtonDialpad primaryText="1" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with primaryText', async () => {
      expect.assertions(1);

      const { container } = await setup(<ButtonDialpad primaryText="1" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with secondaryText', async () => {
      expect.assertions(1);

      const { container } = await setup(<ButtonDialpad secondaryText="ABC" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with primaryText and secondaryText', async () => {
      expect.assertions(1);

      const { container } = await setup(<ButtonDialpad primaryText="1" secondaryText="ABC" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when disabled', async () => {
      expect.assertions(1);

      const { container } = await setup(
        <ButtonDialpad disabled={true} primaryText="Example Text" />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with title', async () => {
      expect.assertions(1);

      const title = 'Example Text';

      const { container } = await setup(<ButtonDialpad title={title} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its main class', async () => {
      expect.assertions(1);

      await setup(<ButtonDialpad primaryText="Example Text" />);
      const element = screen.getByRole('button');

      expect(element.classList.contains(STYLE.wrapper)).toBe(true);
    });

    it('should pass disabled prop', async () => {
      expect.assertions(1);

      await setup(<ButtonDialpad disabled={true} primaryText="Example Text" />);
      const element = screen.getByRole('button');

      expect(element.getAttribute('disabled')).not.toBeNull();
    });

    it('should have custom class if provided', async () => {
      const testClass = 'testClass';

      await setup(<ButtonDialpad className={testClass} primaryText="Example Text" />);
      const element = screen.getByRole('button');

      expect(element.classList.contains(testClass)).toBe(true);
    });

    it('should pass primaryText prop', async () => {
      expect.assertions(1);

      const primaryText = '1';

      await setup(<ButtonDialpad primaryText={primaryText} />);
      const element = screen.getByRole('button');
      const target = element.getElementsByClassName(STYLE.primaryText)[0];

      expect(target.innerHTML).toBe(primaryText);
    });

    it('should pass secondaryText prop', async () => {
      expect.assertions(1);

      const secondaryText = 'ABC';

      await setup(<ButtonDialpad secondaryText={secondaryText} />);
      const element = screen.getByRole('button');
      const target = element.getElementsByClassName(STYLE.secondaryText)[0];

      expect(target.innerHTML).toBe(secondaryText);
    });

    it('should have provided title when title is provided', async () => {
      expect.assertions(1);

      const title = 'Example Text';

      await setup(<ButtonDialpad title={title} primaryText="Example Text" />);
      const element = screen.getByRole('button');

      expect(element.getAttribute('title')).toBe(title);
    });

    it('should render MdcButton', async () => {
      expect.assertions(1);

      const { container } = await setup(<ButtonDialpad primaryText="Example Text" />);
      const element = container.querySelector('mdc-button');

      expect(element).not.toBeNull();
    });
  });

  describe('actions', () => {
    const user = userEvent.setup();

    it('should handle mouse click events', async () => {
      expect.assertions(1);

      const mockCallback = jest.fn();

      await setup(<ButtonDialpad onClick={mockCallback} primaryText="Example Text" />);

      const component = screen.getByRole('button');
      await user.click(component);

      expect(mockCallback).toBeCalledTimes(1);
    });
  });
});
