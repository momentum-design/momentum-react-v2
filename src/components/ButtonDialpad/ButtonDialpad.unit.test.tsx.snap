// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonPill /> snapshot should match snapshot 1`] = `
<div>
  <mdc-button
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    primarytext="1"
    role="button"
    size="40"
    tabindex="0"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      >
        1
      </div>
      <div
        class="md-button-dialpad-secondary-text"
      />
    </div>
  </mdc-button>
</div>
`;

exports[`<ButtonPill /> snapshot should match snapshot when disabled 1`] = `
<div>
  <mdc-button
    aria-disabled="true"
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    disabled=""
    internals-disabled=""
    primarytext="Example Text"
    role="button"
    size="40"
    tabindex="-1"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      >
        Example Text
      </div>
      <div
        class="md-button-dialpad-secondary-text"
      />
    </div>
  </mdc-button>
</div>
`;

exports[`<ButtonPill /> snapshot should match snapshot with primaryText 1`] = `
<div>
  <mdc-button
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    primarytext="1"
    role="button"
    size="40"
    tabindex="0"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      >
        1
      </div>
      <div
        class="md-button-dialpad-secondary-text"
      />
    </div>
  </mdc-button>
</div>
`;

exports[`<ButtonPill /> snapshot should match snapshot with primaryText and secondaryText 1`] = `
<div>
  <mdc-button
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    primarytext="1"
    role="button"
    secondarytext="ABC"
    size="40"
    tabindex="0"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      >
        1
      </div>
      <div
        class="md-button-dialpad-secondary-text"
      >
        ABC
      </div>
    </div>
  </mdc-button>
</div>
`;

exports[`<ButtonPill /> snapshot should match snapshot with secondaryText 1`] = `
<div>
  <mdc-button
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    role="button"
    secondarytext="ABC"
    size="40"
    tabindex="0"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      />
      <div
        class="md-button-dialpad-secondary-text"
      >
        ABC
      </div>
    </div>
  </mdc-button>
</div>
`;

exports[`<ButtonPill /> snapshot should match snapshot with title 1`] = `
<div>
  <mdc-button
    class="md-button-dialpad-wrapper"
    color="default"
    data-btn-type="pill"
    role="button"
    size="40"
    tabindex="0"
    title="Example Text"
    type="button"
    variant="tertiary"
  >
    <div
      class="md-button-dialpad-inner-wrapper"
    >
      <div
        class="md-button-dialpad-primary-text"
      />
      <div
        class="md-button-dialpad-secondary-text"
      />
    </div>
  </mdc-button>
</div>
`;
